File: app/src/main/java/life/plank/juna/zone/view/fragment/forum/ForumFragment.java
package life.plank.juna.zone.view.fragment.forum;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import com.google.gson.JsonObject;
import com.squareup.picasso.Picasso;

import java.net.HttpURLConnection;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

import javax.inject.Inject;
import javax.inject.Named;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import life.plank.juna.zone.R;
import life.plank.juna.zone.data.model.FeedItemComment;
import life.plank.juna.zone.data.network.interfaces.RestApi;
import life.plank.juna.zone.interfaces.FeedInteractionListener;
import life.plank.juna.zone.util.AppConstants;
import life.plank.juna.zone.util.UIDisplayUtil;
import life.plank.juna.zone.view.adapter.post.PostCommentAdapter;
import retrofit2.Response;
import retrofit2.Retrofit;
import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

import static life.plank.juna.zone.ZoneApplication.getApplication;
import static life.plank.juna.zone.util.PreferenceManager.getSharedPrefsString;

public class ForumFragment extends Fragment implements FeedInteractionListener {

    private static String matchBoardId;
    @Inject
    Picasso picasso;
    @Inject
    @Named("default")
    Retrofit retrofit;
    @BindView(R.id.post_comments_list)
    RecyclerView postCommentsRecyclerView;
    @BindView(R.id.post_comment)
    TextView postComment;
    @BindView(R.id.comment_edit_text)
    EditText commentEditText;
    String date;
    private String userId;
    private PostCommentAdapter adapter;
    private RestApi restApi;

    public static ForumFragment newInstance(String boardId) {
        ForumFragment fragment = new ForumFragment();
        matchBoardId = boardId;
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getApplication().getUiComponent().inject(this);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_forum, container, false);
        ButterKnife.bind(this, rootView);
        adapter = new PostCommentAdapter(picasso, this, getString(R.string.forum));
        postCommentsRecyclerView.setAdapter(adapter);
        setAdapterData();
        restApi = retrofit.create(RestApi.class);
        date = new SimpleDateFormat(getString(R.string.string_format)).format(Calendar.getInstance().getTime());
        SharedPreferences preference = UIDisplayUtil.getSignupUserData(getContext());
        userId = preference.getString(getString(R.string.pref_object_id), getString(R.string.na));
        return rootView;
    }

    //TODO: Remove hard coded data after backend integration.
    public void setAdapterData() {

        FeedItemComment firstFeedItemComment = new FeedItemComment("1", getString(R.string.first_feed_comment),
                "ROFootball",
                null, 0, false, 3, null);

        FeedItemComment secondFeedItemComment = new FeedItemComment("2", getString(R.string.first_feed_comment),
                "ROFootball",
                null, 3, true, 3, null);

        List<FeedItemComment> commentList = new ArrayList<>();
        commentList.add(firstFeedItemComment);
        commentList.add(secondFeedItemComment);
        adapter.setComments(commentList);

    }

    @OnClick({R.id.post_comment})
    public void onViewClicked() {
        postCommentOnBoardFeed(commentEditText.getText().toString(), matchBoardId, AppConstants.ROOT_COMMENT, userId, date);
    }

    //TODO: Update url
    private void postCommentOnBoardFeed(String getEditTextValue, String boardId, String contentType, String userId, String dateCreated) {

        String token = getString(R.string.bearer) + " " + getSharedPrefsString(getString(R.string.pref_login_credentails), getString(R.string.pref_azure_token));

        restApi.postFeedItemOnBoard(getEditTextValue, boardId, contentType, userId, dateCreated, token)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Subscriber<Response<JsonObject>>() {
                    @Override
                    public void onCompleted() {
                        Log.i("", "onCompleted: ");
                    }

                    @Override
                    public void onError(Throwable e) {
                        Log.e("", "onError: " + e);
                        Toast.makeText(getContext(), R.string.something_went_wrong, Toast.LENGTH_LONG).show();
                    }

                    @Override
                    public void onNext(Response<JsonObject> jsonObjectResponse) {

                        switch (jsonObjectResponse.code()) {
                            case HttpURLConnection.HTTP_OK:
                                //TODO: Refresh content
                                break;
                            case HttpURLConnection.HTTP_BAD_REQUEST:
                                Toast.makeText(getContext(), R.string.something_went_wrong, Toast.LENGTH_LONG).show();
                                break;
                            default:
                                Toast.makeText(getContext(), R.string.something_went_wrong, Toast.LENGTH_LONG).show();
                                break;
                        }
                    }
                });
    }

    @Override
    public void onCommentLiked() {

    }

    @Override
    public void onCommentDisliked() {

    }

    @Override
    public void onPostCommentOnFeed() {
    }

    @Override
    public void onPostReplyOnComment(String reply, int position, FeedItemComment comment) {

    }
}

File: app/src/main/java/life/plank/juna/zone/view/fragment/forum/ForumFragment.kt
package life.plank.juna.zone.view.fragment.forum

import android.os.Bundle
import android.text.InputType.*
import android.util.Log
import android.view.*
import android.view.inputmethod.EditorInfo
import android.widget.EditText
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.prembros.facilis.util.*
import kotlinx.android.synthetic.main.fragment_forum.*
import life.plank.juna.zone.R
import life.plank.juna.zone.ZoneApplication.getApplication
import life.plank.juna.zone.data.model.FeedItemComment
import life.plank.juna.zone.data.network.interfaces.RestApi
import life.plank.juna.zone.notification.*
import life.plank.juna.zone.util.common.*
import life.plank.juna.zone.util.common.AppConstants.NEW_LINE
import life.plank.juna.zone.util.common.JunaDataUtil.findString
import life.plank.juna.zone.util.sharedpreference.PreferenceManager
import life.plank.juna.zone.util.sharedpreference.PreferenceManager.Auth.getToken
import life.plank.juna.zone.view.controller.ForumCommentController
import life.plank.juna.zone.view.fragment.base.BaseCommentContainerFragment
import java.net.HttpURLConnection.*
import javax.inject.Inject

class ForumFragment : BaseCommentContainerFragment() {

    @Inject
    lateinit var restApi: RestApi

    private var boardId: String? = null
    private var commentList: MutableList<FeedItemComment>? = null
    private var forumCommentController: ForumCommentController? = null
    private var commentId: String? = null
    private var parentCommentId: String? = null

    companion object {
        private val TAG = ForumFragment::class.java.simpleName
        fun newInstance(boardId: String?) = ForumFragment().apply { arguments = Bundle().apply { putString(findString(R.string.intent_board_id), boardId) } }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        getApplication().uiComponent.inject(this)
        arguments?.run { boardId = getString(getString(R.string.intent_board_id)) }
        commentId = activity?.getChildIdFromIntent()
        parentCommentId = activity?.getSiblingIdFromIntent()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? = inflater.inflate(R.layout.fragment_forum, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        if (boardId == null) return

        prepareList()
        prepareViews()
        setListeners()
    }

    private fun prepareList() {
        forumCommentController = ForumCommentController(this)
        post_comments_list.adapter = forumCommentController?.adapter

        forumCommentController?.addModelBuildListener {
            if (!isNullOrEmpty(commentId)) {
                commentList?.run {
                    if (isNullOrEmpty(parentCommentId)) {
//                        Scroll to comment
                        var commentIndex = indexOf(find { comment -> comment.id == commentId })
                        if (commentIndex == -1) commentIndex = 0
                        scrollToComment(commentIndex)
                    } else {
//                        Scroll to reply
                        val comment = find { comment -> comment.id == parentCommentId }
                        var parentCommentIndex = indexOf(comment)
                        if (parentCommentIndex == -1) parentCommentIndex = 0

                        val replyIndex = comment?.replies?.indexOf(comment.replies?.find { reply -> reply.id == commentId })
                                ?: -1
                        if (replyIndex == -1) {
                            scrollToComment(parentCommentIndex)
                        } else {
                            scrollToReply(parentCommentIndex, replyIndex)
                        }
                        parentCommentId = null
                        activity?.removeIntentExtra(R.string.intent_sibling_id)
                    }
                }
                commentId = null
                activity?.removeIntentExtra(R.string.intent_child_id)
            }
        }
    }

    private fun prepareViews() {
        Glide.with(this)
                .load(PreferenceManager.CurrentUser.getProfilePicUrl())
                .into(commenter_image)

        if (PreferenceManager.App.isEnterToSend()) {
            comment_edit_text.inputType = TYPE_CLASS_TEXT or TYPE_TEXT_FLAG_CAP_SENTENCES
            comment_edit_text.imeOptions = EditorInfo.IME_ACTION_SEND
        } else {
            comment_edit_text.inputType = TYPE_CLASS_TEXT or TYPE_TEXT_FLAG_MULTI_LINE or TYPE_TEXT_FLAG_CAP_SENTENCES
            comment_edit_text.imeOptions = EditorInfo.IME_ACTION_NONE
        }
    }

    private fun setListeners() {
        post_comment.onDebouncingClick {
            if (!comment_edit_text.text.toString().trim().isEmpty() && !isNullOrEmpty(boardId)
                    && !comment_edit_text.text.toString().matches("@[a-zA-Z0-9]+\\s*".toRegex())) {
                post_comment.clearFocus()
                postCommentOrReply(comment_edit_text.text.toString(), getCommentEventForBoardComment(boardId!!))
            }

        }
        forum_swipe_refresh_layout.setOnRefreshListener { getComments(true) }

        comment_edit_text.setOnEditorActionListener { _, actionId, _ ->
            return@setOnEditorActionListener when (actionId) {
                EditorInfo.IME_ACTION_SEND -> {
                    if (!comment_edit_text.text.toString().trim().isEmpty() && !isNullOrEmpty(boardId)
                            && !comment_edit_text.text.toString().matches("@[a-zA-Z0-9]+\\s*".toRegex())) {
                        post_comment.clearFocus()
                        postCommentOrReply(comment_edit_text.text.toString(), getCommentEventForBoardComment(boardId!!))
                    }
                    true
                }
                EditorInfo.IME_ACTION_NONE -> {
                    comment_edit_text.append(NEW_LINE)
                    true
                }
                else -> false
            }
        }
    }

    override fun onResume() {
        super.onResume()
        getComments(false)
    }

    fun getComments(isRefreshing: Boolean) {
        no_comment_text_view.visibility = View.GONE
        restApi.getCommentsForBoard(boardId, getToken())
                .onTerminate { if (isRefreshing) forum_swipe_refresh_layout.isRefreshing = false }
                .setObserverThreadsAndSmartSubscribe({
                    Log.e(TAG, "getComments()", it)
                    errorToast(R.string.failed_to_get_feed_comments, it)
                }, {
                    when (it.code()) {
                        HTTP_OK -> {
                            commentList = it.body()
//                            TODO: pass previousPageToken and nextPageToken when implemented
                            forumCommentController?.setData(commentList, null, null)
                        }
                        HTTP_NOT_FOUND, HTTP_NO_CONTENT -> {
                            no_comment_text_view.setText(R.string.be_the_first_to_comment_on_this_forum)
                            no_comment_text_view.visibility = View.VISIBLE
                            no_comment_text_view.clearOnClickListener()
                        }
                        else -> {
                            no_comment_text_view.setText(R.string.failed_to_get_feed_comments_tap_to_retry)
                            no_comment_text_view.visibility = View.VISIBLE
                            errorToast(R.string.failed_to_get_feed_comments, it)
                            no_comment_text_view.onDebouncingClick { getComments(false) }
                        }
                    }
                }, this)
    }

    override fun specifyCommentEvent() {
        boardId?.run {
            commentEvent = getCommentEventForBoardComment(this)
        }
    }

    override fun commentsRecyclerView(): RecyclerView = post_comments_list

    override fun getTheRestApi(): RestApi = restApi

    override fun getCommentEditText(): EditText = comment_edit_text

    override fun onPostReplyOnComment(reply: String, position: Int, parentComment: FeedItemComment) {
        boardId?.run {
            postReplyOnBoardComment(reply, getCommentEventForReply(this, parentComment.id), parentComment, position)
        }
    }

    override fun onCommentSuccessful(responseComment: FeedItemComment) {
        comment_edit_text.text = null
        no_comment_text_view.visibility = View.GONE
        commentList?.add(0, responseComment)
//        TODO: pass previousPageToken and nextPageToken when implemented
        forumCommentController?.setData(commentList, null, null)
        scrollToComment(0)
    }

    override fun onReplySuccessful(responseReply: FeedItemComment, parentComment: FeedItemComment?, parentCommentPosition: Int, replyPosition: Int) {
        comment_edit_text.text = null
        commentList?.get(parentCommentPosition)?.run {
            if (replies == null) {
                replies = ArrayList()
            }
            (replies as ArrayList).add(responseReply)
            replyCount++
        }
//        TODO: pass previousPageToken and nextPageToken when implemented
        forumCommentController?.setData(commentList, null, null)
        scrollToReply(parentCommentPosition, replyPosition)
    }
}